Builder-Webpack5: Categorize builder error

yannbf opened the pull request:
Closes #24030

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->


## What I did

Webpack builder errors before were just an object with a `compilation` property.
This PR revamps both the `start` and `build` processes in the Webpack5 builder, adding categorized Webpack errors, turning Webpack compilation errors into a single, composite error.

Turns out there was a bug since Storybook 7 where, if you have Storybook successfully running on dev mode, but then introduce a compilation error, e.g. change the code in `Button.js` to have incorrect syntax like `conzt foo` instead of `const foo`, the error would not be propagated to the terminal anymore. This was the case in Storybook 6.5, and it stopped working because of changes in the webpack-dev-middleware library. This PR brings that behavior back.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

Please check unit tests

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version [`0.0.0-pr-24031-sha-553789e8`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24031-sha-553789e8). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24031-sha-553789e8`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24031-sha-553789e8) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`yann/webpack-builder-error`](https://github.com/storybookjs/storybook/tree/yann/webpack-builder-error) |
| **Commit** | [`553789e8`](https://github.com/storybookjs/storybook/commit/553789e83af83c57179659c26223d87f043f8460) |
| **Datetime** | Sun Sep  3 21:15:01 UTC 2023 (`1693775701`) |
| **Workflow run** | [6066789759](https://github.com/storybookjs/storybook/actions/runs/6066789759) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24031`_
</details>
<!-- CANARY_RELEASE_SECTION -->

shilman commented on code/lib/core-events/src/errors/server-errors.test.ts:
```diff
@@ -0,0 +1,61 @@
+
+    const webpackError = new WebpackCompilationError(data);
+
+    const expectedErrorMessage = compilationErrors
```
Can you hard code this instead? Hard to read what the result is supposed to be in the current form

yannbf replied:
Done!

shilman reviewed (approved):
Nice one! 💪

yannbf committed:
treat empty error messages in telemetry reports and allow non-error instances

tmeasday commented on code/lib/core-events/src/errors/server-errors.test.ts:
```diff
@@ -0,0 +1,61 @@
+
+  it('should correctly handle error with error within error', () => {
+    const error = new Error('Custom error message');
+    const data = {
+      error: new Error(),
+    };
+    data.error = error;
```
Is this supposed to be `data.error.error = error`?

yannbf replied:
great catch! fixed

tmeasday commented on code/lib/core-events/src/errors/server-errors.ts:
```diff
@@ -137,3 +137,48 @@ export class InvalidStoriesEntryError extends StorybookError {
+      } else {
+        this.errorMessage = data.error.message;
+      }
+    } else if (data.error.compilation?.errors) {
+      data.error.compilation.errors.forEach((e: Error) => {
+        this.errorMessage += `${e.name}: ${e.message}\n\n`;
+      });
+    }
```
We don't have a test for this bit?

yannbf replied:
great catch! done!

yannbf committed:
categorize webpack builder error

ndelangen commented on code/builders/builder-webpack5/src/index.ts:
```diff
@@ -195,8 +196,7 @@ const starter: StarterFunction = async function* starterGeneratorFn({
   if (stats.hasErrors()) {
-    // eslint-disable-next-line @typescript-eslint/no-throw-literal
-    throw stats;
+    throw new WebpackCompilationError({ error: stats });
```
Don't merge this suggested code, it's just a starting point for the convo:

It seems weird to me that we'd lump these different types of issues together.
Should we separate these into multiple different  
```suggestion
    throw new WebpackCompilationStatsError({ stats });
```

Or we separate the different input types like so:
```suggestion
    throw new WebpackCompilationError({ stats });
```

WDTY?

yannbf committed:
standardize errors thrown in webpack builder

- also bring back error logging on dev mode, which seemed to be lost in Storybook 7.0.0

yannbf committed:
cleanup tests

tmeasday reviewed (approved):
LGTM!

yannbf committed:
Merge branch 'next' into yann/webpack-builder-error
